### ISSUE TITLE
Reference table (or subquery) left join distributed table error out

### ISSUE THREAD

**Issue by colton-citus** [citusdata/citus#2614]
Messages about the following are seen with certain subqueries and CTEs:
```
ERROR:  cannot pushdown the subquery
DETAIL:  There exist a reference table in the outer part of the outer join
```

This behavior can only be reliably reproduced so far when both of the following conditions are met:
1. There are LEFT JOINs between 2 different reference tables
2. There is a subquery or CTE of any sort in the query

The following minimum reproduction script contains multiple queries that each trigger this behavior: 

```
drop table if exists distributed_1;
drop table if exists distributed_2;
drop table if exists reference_1;
drop table if exists reference_2;

create table distributed_1 (col1 int, col2 int, distrib_col int);
create table distributed_2 (col1 int, col2 int, distrib_col int);
create table reference_1 (col1 int, col2 int);
create table reference_2(col1 int, col2 int);


select create_distributed_table('distributed_1','distrib_col');
select create_distributed_table('distributed_2','distrib_col');
select create_reference_table('reference_1');
select create_reference_table('reference_2');



select r1.* from distributed_1 AS d1
LEFT JOIN reference_1 AS r1 ON d1.col2=r1.col2
LEFT JOIN reference_2 AS r2 ON r2.col1 = r1.col1
join (select distrib_col,count(*) from distributed_2 group by distrib_col) d2 ON d2.distrib_col=d1.distrib_col
;

--If there's no aggregate in the join, it complains about subqueries instead


with d2 AS (select distrib_col,count(*) from distributed_2 group by distrib_col)
select r1.* from distributed_1 AS d1
LEFT JOIN reference_1 AS r1 ON d1.col2=r1.col2
LEFT JOIN reference_2 AS r2 ON r2.col1 = r1.col1
join d2 ON d2.distrib_col=d1.distrib_col
;

with d2 AS (select distrib_col,col1 from distributed_2)
select r1.* from distributed_1 AS d1
LEFT JOIN reference_1 AS r1 ON d1.col2=r1.col2
LEFT JOIN reference_2 AS r2 ON r2.col1 = r1.col1
join d2 ON d2.distrib_col=d1.distrib_col
;

with cte_1 AS (select col1 from reference_1)
select r1.* from distributed_1 AS d1
LEFT JOIN reference_1 AS r1 ON d1.col2=r1.col2
LEFT JOIN reference_2 AS r2 ON r2.col1 = r1.col1
join cte_1 ON cte_1.col1=d1.distrib_col
;

```

**Comment by colton-citus** [comment]
@saicitus and @samay-sharma have both reported seeing similar behavior with customer queries.

**Comment by onderkalaci** [comment]
@colton-citus @saicitus and @samay-sharma, these queries are now working on master via #2481. So, 8.2 will include this improvement.

And, for the readers of the code, this [line](https://github.com/citusdata/citus/pull/2481/files#diff-659a9573986a07963c9b7bdcaf12fac3R689) has fixed the issue in essence.

Although #2481 fixed this case, there are still plenty of ` Complex subqueries and CTEs cannot be in the outer part of the outer join` errors you could get such as:
```SQL
WITH cte_1 AS (select distrib_col,count(*) from distributed_2 group by distrib_col)
SELECT * FROM cte_1 LEFT JOIN distributed_2 ON (cte_1.distrib_col = distributed_2.distrib_col);
ERROR:  cannot pushdown the subquery
DETAIL:  Complex subqueries and CTEs cannot be in the outer part of the outer join
``` 

For those cases (and some more), we've a prototype #2504 which solves them, though using pull&push execution. 

So, my question would be how often do you see `Complex subqueries and CTEs cannot be in the outer part of the outer join` or ` complex joins are only supported when all distributed tables are joined on their distribution columns with equal operator`?

**Comment by saicitus** [comment]
@onderkalaci we see the `Complex subqueries and CTEs cannot be in the outer part of the outer join`  issue quite commonly. 

So in the real-time executor sql coverage issues we see the below errors very commonly (50-60% of the times). Each of them comes up with 50-50 chance.

```
ERROR:  cannot pushdown the subquery
DETAIL:  Complex subqueries and CTEs cannot be in the outer part of the outer join
```
```
ERROR:  cannot pushdown the subquery
DETAIL:  There exist a reference table in the outer part of the outer join
```

**Comment by saimicrosoft** [comment]
@onderkalaci +1. Customer ran into this in a technical session today. Overall SQL coverage has improved tremendously but still this error comes up often

**Comment by zlepper** [comment]
Looks like I'm encountering an issue that looks somewhat like this. Specifically we use EntityFramework to generate the querries, and it's currently generating a query like this:

```sql
SELECT t0."Id",
       l."TenantId",
       l."Id",
       m."TenantId",
       m."Id",
       m."LanguageId",
       m."MetaFieldId",
       m0."TenantId",
       m0."Id",
       m0."Name"
FROM (
         SELECT t."Id"
         FROM "Tenants" AS t
         WHERE t."Id" = 'eafbe893-008f-4864-897b-ffb9b30ee263'
         LIMIT 1
     ) AS t0
         LEFT JOIN "Languages" AS l ON t0."Id" = l."TenantId"
         LEFT JOIN "MetaFieldLabels" AS m ON t0."Id" = m."TenantId"
         LEFT JOIN "MetaFields" AS m0 ON t0."Id" = m0."TenantId"
ORDER BY t0."Id", l."TenantId", l."Id", m."TenantId", m."Id", m0."TenantId", m0."Id"
```

which then fails with ` ERROR: complex joins are only supported when all distributed tables are co-located and joined on their distribution columns`

Just asking for the entire list (Which is still only one tenant due to the `where` clause), and then just grapping the first one client side gets around the issue for now. 

Working:
```c#
return (await Tenants
                .Include(t => t.Languages)
                .Include(l => l.MetaFieldLabels)
                .Include(t => t.MetaFields)
                .Where(t => t.Id == tenantId)
                .ToListAsync(cancellationToken))[0];
```

Not working:

```c#
return await Tenants
                .Include(t => t.Languages)
                .Include(l => l.MetaFieldLabels)
                .Include(t => t.MetaFields)
                .FirstAsync(t => t.Id == tenantId, cancellationToken);
```

**Comment by marcocitus** [comment]
Fixed by #6512
